### ISSUE TITLE
graphql: custom fields not accessible to userland resolvers

### ISSUE THREAD

**Issue by PaulieScanlon** [gatsbyjs/gatsby#30429]
## Description

Hi team, i'm trying to create a nested set of routes from articles sourced from Shopify and it looks like some of the curly braces magic isn't quite working as i'd hoped 

### Steps to reproduce

This is my dir structure on disk

```sh
src
   |-- pages
      |-- {shopifyArticle.blog__title} // => this is a dir not a file FYI
         |-- {shopifyArticle.title}.tsx
```

This will result in pages being correctly created and can be seen on the "Gatsby.js development 404 page" and by visiting the url. So... the pages are definitely there. 

However from a page query or a `useStaticQuery` When i attempt to create the `articlePath` using `gatsbyPath` i run into problems.

```javascript
import { graphql, useStaticQuery } from 'gatsby'

export const useShopifyArticle = () => {
  return useStaticQuery(graphql`
    query {
      allShopifyArticle {
        nodes {
          title
          articlePath: gatsbyPath(filePath: "/{shopifyArticle.blog__title}/{shopifyArticle.title}")
        }
      }
    }
  `)
}
```

### Expected result

```javascript
{
  "allShopifyArticle": {
    "nodes": [
      {
        "title": "Article Title",
        "articlePath": "/blog-title/article-title/"
      }
    ]
  }
}
```

### Actual result

```javascript
{
  "allShopifyArticle": {
    "nodes": [
      {
        "title": "Article Title",
        "articlePath": "/{shopifyArticle.blog__title}/article-title/"
      }
    ]
  }
}
```

### Environment

  System:
    OS: macOS Mojave 10.14.5
    CPU: (8) x64 Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.16.1 - /usr/local/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.13.4 - /usr/local/bin/npm
  Languages:
    Python: 2.7.10 - /usr/bin/python
  Browsers:
    Chrome: 89.0.4389.90
    Firefox: 86.0.1
    Safari: 13.1.1
  npmPackages:
    gatsby: ^3.1.1 => 3.1.1 
    gatsby-plugin-gatsby-cloud: ^2.1.0 => 2.1.0 
    gatsby-plugin-image: ^1.1.1 => 1.1.1 
    gatsby-plugin-sharp: ^3.1.1 => 3.1.1 
    gatsby-plugin-theme-ui: ^0.6.0 => 0.6.0 
    gatsby-source-shopify: ^3.10.1 => 3.10.1 
    gatsby-transformer-sharp: ^3.1.0 => 3.1.0 
  npmGlobalPackages:
    gatsby-cli: 2.15.1

**Comment by LekoArts** [comment]
Hi!

Can you create a [minimal reproduction][reproduction] without Shopify for this? Because if it's broken, it should also be broken with e.g. MDX. But on the Shopify test project I have at hand it works here:

https://github.com/gatsbyjs/gatsby-starter-shopify/blob/dd3735ebf6c5cf7d4b494d3fc5b31d7ab5a47519/src/components/product-card.jsx#L74-L76

[reproduction]: https://gatsby.dev/reproduction

**Comment by PaulieScanlon** [comment]
Hey @LekoArts thanks for looking into this!! Yeap sure thing. I was thinking the same thing. Gimmie an hour and i'll spin up an MDX version and see if the problem persists

Ty!

**Comment by PaulieScanlon** [comment]
@LekoArts Hey mate. Here's a [reproduction repo](https://github.com/PaulieScanlon/file-system-routes-gatsby-path)

I still "kind of" see the same issue only this time it's the second part of the URL that doesn't escape: `{mdx.slug}`. 
Where as in the Shopify example it's the first part of the URL that doesn't escape: `{shopifyArticle.blog__title}`

Not sure if this is an MDX issue or if it confuses the Shopify issue i original experienced but in either case it would appear as though it's not possible to correctly construct an `articlePath` despite the pages actually being created correctly.

```javascript
{
  allMdx {
    nodes {
      frontmatter {
        title
        blog
      }
      slug
      articlePath: gatsbyPath(filePath: "/{mdx.frontmatter__blog}/{mdx.slug}")
    }
  }
}

```

## Expected result

```javascript
  "data": {
    "allMdx": {
      "nodes": [
        {
          "frontmatter": {
            "title": "Article 2",
            "blog": "cats"
          },
          "slug": "article-2",
          "articlePath": "/cats/article-1/"
        },
      ]
    }
  }
```

## Actual Result

```javascript
  "data": {
    "allMdx": {
      "nodes": [
        {
          "frontmatter": {
            "title": "Article 2",
            "blog": "cats"
          },
          "slug": "article-2",
          "articlePath": "/cats/{mdx/"
        },
      ]
    }
  }
```

**Comment by PaulieScanlon** [comment]
> Hi!
> 
> Can you create a [minimal reproduction](https://gatsby.dev/reproduction) without Shopify for this? Because if it's broken, it should also be broken with e.g. MDX. But on the Shopify test project I have at hand it works here:
> 
> https://github.com/gatsbyjs/gatsby-starter-shopify/blob/dd3735ebf6c5cf7d4b494d3fc5b31d7ab5a47519/src/components/product-card.jsx#L74-L76

Just out of curiosity what happens if you remove the hardcoded "/products" at the beginning of the filePath: https://github.com/gatsbyjs/gatsby-starter-shopify/blob/dd3735ebf6c5cf7d4b494d3fc5b31d7ab5a47519/src/components/product-card.jsx#L75

I'm wondering if there needs to be some kind of string in the file path for it to work properly?

**Comment by LekoArts** [comment]
So the issue is that `mdx.slug` isn't available when we create that path:

![image](https://user-images.githubusercontent.com/16143594/112305611-331a9180-8c9f-11eb-94a4-a0c415b1ba73.png)

I'll look into it why that is

**Comment by PaulieScanlon** [comment]
@LekoArts lovely stuff! Thanks mate.

Is this also the reason `{shopifyArticle.blog__title}` isn't available too? or is that a Shopify / Shopify Source issue?

**Comment by LekoArts** [comment]
I should rephrase: `mdx.slug` is there in the GraphQL query, but in this particular step (the raw node) not. That is maybe also happening with Shopify and how it constructs its nodes.

It would be of great help if you could go into your `node_modules` and add a `console.log({ sourceCopy })` here: https://github.com/gatsbyjs/gatsby/blob/823140f2b0bbbcab51923186bab8128bb8e0afe5/packages/gatsby-plugin-page-creator/src/gatsby-node.ts#L198 and paste the result here. I want to see if `blog { title }` is there

**Comment by PaulieScanlon** [comment]
@LekoArts Hey mate!

## Reproduction Repo
This is from the reproduction repo

```javascript
{
sourceCopy: {
    id: 'c6c61a07-1ff9-574d-b64c-de8478e0a5e3',
    children: [],
    parent: {
      id: '3bf48e94-5b9f-5195-b4b6-85957b61cfcc',
      children: [Array],
      parent: null,
      internal: [Object],
      sourceInstanceName: '__PROGRAMMATIC__',
      relativePath: 'article-4.mdx',
      extension: 'mdx',
      prettySize: '60 B',
      modifiedTime: '2021-03-24T09:50:59.501Z',
      accessTime: '2021-03-24T14:42:29.690Z',
      changeTime: '2021-03-24T09:50:59.501Z',
      birthTime: '2021-03-24T09:30:08.264Z',
      root: '/',
      dir: '/Users/paulscanlon/Desktop/_dev-issues/file-system-routes-gatsby-path/src/content',
      base: 'article-4.mdx',
      ext: '.mdx',
      name: 'article-4',
      absolutePath:
'/Users/paulscanlon/Desktop/_dev-issues/file-system-routes-gatsby-path/src/content/article-4.mdx',
      relativeDirectory: '',
      dev: 16777220,
      mode: 33188,
      nlink: 1,
      uid: 501,
      rdev: 0,
      blksize: 4096,
      ino: 68592716,
      size: 60,
      blocks: 8,
      atimeMs: 1616596949690.2112,
      mtimeMs: 1616579459500.8062,
      ctimeMs: 1616579459500.8062,
      birthtimeMs: 1616578208264.463,
      atime: '2021-03-24T14:42:29.690Z',
      mtime: '2021-03-24T09:50:59.501Z',
      ctime: '2021-03-24T09:50:59.501Z',
      birthtime: '2021-03-24T09:30:08.264Z'
    },
    internal: {
      content: '---\ntitle: Article 4\nblog: dogs\n---\n\nThis is some body text\n',
      type: 'Mdx',
      contentDigest: '323edf6f59921c6da7c8c1589995fd86',
      counter: 28,
      owner: 'gatsby-plugin-mdx'
    },
    frontmatter: { title: 'Article 4', blog: 'dogs' },
    excerpt: undefined,
    exports: {},
    rawBody: '---\ntitle: Article 4\nblog: dogs\n---\n\nThis is some body text\n',
    fileAbsolutePath:
'/Users/paulscanlon/Desktop/_dev-issues/file-system-routes-gatsby-path/src/content/article-4.mdx'
  }
}
```

## Shopify
This is from my WIP Shopify project

```javascript
{
  sourceCopy: {
    author: {
      bio: null,
      email: '...',
      firstName: 'Paul',
      lastName: 'Scanlon',
      name: 'Paul Scanlon'
    },
    blog: { id: 'Z2lkOi8vc2hvcGlmeS9CbG9nLzc3ODQ3MDY4ODQ2' },
    comments: { edges: [] },
    content: 'A blog post about blah',
    contentHtml: 'A blog post about blah',
    excerpt: '',
    excerptHtml: '',
    handle: 'blah',
    image: {
      altText: '',
      id: 'Z2lkOi8vc2hvcGlmeS9BcnRpY2xlSW1hZ2UvMTAxNjY5ODYzNTk4',
      src:
'https://cdn.shopify.com/s/files/1/0555/2819/4222/articles/sj_facc057d-30c9-49c3-b87a-4c0d58c89f17.jpg?v=1616437275',
      localFile___NODE: '1fb51416-d6f0-53f9-9457-9ed3f079d37a'
    },
    publishedAt: '2021-03-22T18:21:14Z',
    tags: [],
    title: 'blah',
    url: 'https://xxxx-xxxx.myshopify.com/blogs/test-blog/blah',
    seo: { title: null, description: null },
    shopifyId: 'Z2lkOi8vc2hvcGlmeS9BcnRpY2xlLzU1NTYxMTM1NzM1OA==',
    id: 'Shopify__Article__Z2lkOi8vc2hvcGlmeS9BcnRpY2xlLzU1NTYxMTM1NzM1OA==',
    parent: undefined,
    children: [],
    internal: {
      type: 'ShopifyArticle',
      contentDigest: 'b6b8df737e2ac597eaccaedb3a0d5197',
      counter: 42,
      owner: 'gatsby-source-shopify'
    },
    blog___NODE: 'Shopify__Blog__Z2lkOi8vc2hvcGlmeS9CbG9nLzc3ODQ3MDY4ODQ2',
    comments___NODE: []
  }
}

```

**Comment by PaulieScanlon** [comment]
I just had a thought... i'm using the default queries for the Articles: https://www.gatsbyjs.com/plugins/gatsby-source-shopify/#shopifyqueries-defaults which doesn't look like it includes the blog title

```javascript
// default query

blog {
   id
}
```

Lemme run a quick test and add the blog title in after the `id` and see if that helps... brb

**Comment by PaulieScanlon** [comment]
Son of a 🤬🤬🤬🤬 ... that was it!

The default query doesn't include `blog.title` by adding it to the `shopifyQueries.articles` it now works

```javascript
// gatsby-config

module.exports = {
  plugins: [
    ...
    {
      resolve: `gatsby-source-shopify`,
      options: {
        shopName: process.env.GATSBY_SHOPIFY_SHOP_NAME,
        accessToken: process.env.GATSBY_SHOPIFY_ACCESS_TOKEN,
        apiVersion: `2021-01`,
        downloadImages: true,
        shopifyQueries: {
          articles: `
          query GetArticles($first: Int!, $after: String) {
            articles(first: $first, after: $after) {

              edges {
                cursor
                node {
                  blog {
                    id
                    title // => added this
                  }
                }
              }
            }
          }
        `,
        },
      },
    },
    ...
  ],
}

```

**Comment by KyleAMathews** [comment]
`slug` isn't on the field — it's a custom field added here https://github.com/gatsbyjs/gatsby/pull/25767

FS routes doesn't work with custom defined fields? That surprises me.

**Comment by alesma** [comment]
Hey, I am having the same issue but using gatsby-source-strapi. I am trying to debug it and added a 

console.log(filePath, sourceCopy); 

and I don't see it called at all, any hints on how to debug this correctly?

Thanks

**Comment by alesma** [comment]
I think i have found the issue, in gatsby-source-strapi the node with the relation data is deleted

https://github.com/strapi/gatsby-source-strapi/blob/master/src/normalize.js#L212

**Comment by alesma** [comment]
https://github.com/strapi/gatsby-source-strapi/issues/329

opened an issue on gatsby-source-strapi repo

**Comment by LekoArts** [comment]
Closing in favor of https://github.com/gatsbyjs/gatsby/issues/37103
